### PR TITLE
TS migration. #2288.

### PR DESCRIPTION
**Pull request by MillenniumFalconMechanic**
#### Reviewers
**Functional:** 
@colinmegill, @seve, @tihuan

---

#### Changes
1. Added Typescript and related packages to dev dependencies.
2. Configured tsconfig, webpack, Babel, lint-staged.
3. Updated eslint to `@typescript-eslint` parser and `airbnb-typescript` config (`plugin:@typescript-eslint/recommended` omitted for now).
3. Migrated JS in `src` and `__tests__` directories to TS.
4. Fixed failing unit tests for runtime errors introduced by `ts-migrate`.

### CODE REVIEW

File: client/configuration/webpack/webpack.config.shared.js [hunk 1]
```diff
@@ -29,6 +29,9 @@ module.exports = {
     path: path.resolve("build"),
     publicPath,
   },
+  resolve: {
+    extensions: [".ts", ".tsx", "..."],
```
Comment (tihuan) [inline client/configuration/webpack/webpack.config.shared.js:33]: for my own education, what's the `"..."` for? Thank you!

**Review by tihuan — COMMENTED** [review]
LGTM! Just one tiny question 😄 👌 Thanks so much again for migrating the code base to TS 🙏 This is super exciting!

**Comment by tihuan** [comment]
Hi @signechambers1 ! We believe the branch is ready for feature testing now 🙏 Could you please help us out here to make sure the app works as expected before we merge this into `main`? 

Branch: `mim/ts-migrate-1-6`

Thanks so much!

File: client/configuration/webpack/webpack.config.shared.js (hunk 1 above)
Comment (MillenniumFalconMechanic) [reply client/configuration/webpack/webpack.config.shared.js:33]: `"..."` includes the default webpack values for the `resolve.extensions` array ([docs](https://webpack.js.org/configuration/resolve/#resolveextensions)).

The `resolve.extensions` config option wasn't specified before I added the `.ts` and `.tsx` values; my intention here was to ensure any default values previously being used (e.g. `.js`, `.jsx` or `.css`) would still be supported. Let me know if you would you prefer an explicit list here!

**Comment by signechambers1** [comment]
LGTM! I tested the following functionality on the krasnow smart seq 2 dataset and everything looked good to me:

Create gene set with single gene
Create gene set with bulk add
Create gene set with gene not in the dataset
Create gene with same name as existing gene set
Rename a gene set
Add a gene to a gene set
Add a duplicate gene to a gene set
Add a gene not in the dataset to a gene set
Remove a gene from a gene set
Delete gene set
Expand gene set
Expand gene
Brushing on histogram
Plotting a gene on x and y
Color by gene in gene set
Color by mean expression
Toggle mean expression
Undo / Redo for create new gene set
Undo / Redo for add new gene
Undo / Redo for rename a gene set
Undo / Redo for delete gene
Undo / Redo for delete gene set
Everything loads
Color by categorical
Pan
Zoom
Lasso
Subset
Create new category
Create new label in category
Delete label
Delete category
Rename label
Rename category
Undo / redo rename category
Clip
Diff exp (includes select via lasso)
Select via histogram
Change embedding
While in subset, color by gene set mean expression
While in subset, color by gene in gene set
While in subset, select via histogram of a single gene
While in subset, select via gene set histogram
Run diffexp on 2 arbitrary cell populations
Quick Gene Search - add gene
Quick Gene Search - color by gene
Quick Gene Search - remove gene
Quick Gene Search - expand gene
Quick Gene Search - select on gene histogram
Quick Gene Search - plot on x and y
Collapse and expand Quick gene search
Collapse and expand gene sets

**Comment by tihuan** [comment]
YAY!! You're the best, @signechambers1 !! Really appreciate you going through all the test cases patiently🙏 
Adding my LGTM now too!

**Review by tihuan — APPROVED** [review]
Whoop whoop! GO TEAM!!

File: client/configuration/webpack/webpack.config.shared.js (hunk 1 above)
Comment (tihuan) [reply client/configuration/webpack/webpack.config.shared.js:33]: Ooh didn't know that! So convenient 😆 Thanks for sharing the knowledge 💡 `"..."` sounds great!

**Review by colinmegill — APPROVED** [review]

**Review by seve — APPROVED** [review]
Full steam ahead!

**Comment by tihuan** [comment]
Whoop whoop GO TEAM!! Thanks all!! 🤩 👏 🚀 🚀 🚀 🚀 🚀

**Comment by tihuan** [comment]
@MillenniumFalconMechanic  I enabled auto merge 🙆‍♂️ Thank you!

**Pull request merged**